File: Novels/packages/5-mobile-ios/StoryReader/Repository/DiscoveryRepository.swift
import Foundation
import Combine

// Discovery Repository - Rankings, Editor Picks, Genre Browsing
class DiscoveryRepository {
    private let graphQLService = GraphQLService.shared
    private let cache = InMemoryCache<CacheKey, [Story]>(expirationInterval: 60 * 10) // 10 minutes
    
    // Get rankings
    func getRankings(rankingType: RankingType, genre: String? = nil, timeRange: TimeRange? = nil) -> AnyPublisher<[Story], Error> {
        let key = CacheKey.rankings(
            type: rankingType.rawValue,
            genre: genre,
            timeRange: timeRange?.rawValue
        )
        
        if let cached = cache.value(forKey: key) {
            return Just(cached)
                .setFailureType(to: Error.self)
                .eraseToAnyPublisher()
        }
        
        return graphQLService.getRankings(
            rankingType: rankingType.rawValue,
            genre: genre,
            timeRange: timeRange?.rawValue
        )
        .handleEvents(receiveOutput: { [weak self] stories in
            self?.cache.insert(stories, forKey: key)
        })
        .eraseToAnyPublisher()
    }
    
    // Get editor picks
    func getEditorPicks(limit: Int? = nil, genre: String? = nil) -> AnyPublisher<[Story], Error> {
        let key = CacheKey.editorPicks(limit: limit, genre: genre)
        
        if let cached = cache.value(forKey: key) {
            return Just(cached)
                .setFailureType(to: Error.self)
                .eraseToAnyPublisher()
        }
        
        return graphQLService.getEditorPicks(limit: limit, genre: genre)
            .handleEvents(receiveOutput: { [weak self] stories in
                self?.cache.insert(stories, forKey: key)
            })
            .eraseToAnyPublisher()
    }
    
    // Get genre stories
    func getGenreStories(genre: String?, page: Int? = nil, limit: Int? = nil) -> AnyPublisher<[Story], Error> {
        let key = CacheKey.genreStories(genre: genre, page: page, limit: limit)
        
        if let cached = cache.value(forKey: key) {
            return Just(cached)
                .setFailureType(to: Error.self)
                .eraseToAnyPublisher()
        }
        
        return graphQLService.getGenreStories(genre: genre, page: page, limit: limit)
            .handleEvents(receiveOutput: { [weak self] stories in
                self?.cache.insert(stories, forKey: key)
            })
            .eraseToAnyPublisher()
    }
    
    enum RankingType: String, Codable {
        case monthlyVotes = "monthly_votes"
        case sales = "sales"
        case recommendations = "recommendations"
        case popularity = "popularity"
    }
    
    enum TimeRange: String, Codable {
        case daily
        case weekly
        case monthly
        case allTime = "all_time"
    }
    
    enum RepositoryError: Error {
        case unknown
        case networkError
    }
    
    private enum CacheKey: Hashable {
        case rankings(type: String, genre: String?, timeRange: String?)
        case editorPicks(limit: Int?, genre: String?)
        case genreStories(genre: String?, page: Int?, limit: Int?)
    }
}

File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/repository/DiscoveryRepository.kt
package com.storysphere.storyreader.repository

import com.storysphere.storyreader.database.AppDatabase
import com.storysphere.storyreader.model.Story
import com.storysphere.storyreader.network.GraphQLService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DiscoveryRepository @Inject constructor(
    private val database: AppDatabase,
    private val graphQLService: GraphQLService
) {
    private val backgroundScope = CoroutineScope(SupervisorJob())
    
    fun getRankings(rankingType: RankingType, genre: String? = null, timeRange: String? = null): Flow<List<Story>> {
        return database.storyDao().getRankings(rankingType.name, genre, timeRange)
            .map { entities -> entities.map { it.toModel() } }
            .onStart {
                backgroundScope.launch {
                    // TODO: Add getRankings to GraphQLService
                    // graphQLService.getRankings(rankingType, genre, timeRange).fold(...)
                }
            }
    }
    
    fun getEditorPicks(limit: Int = 20, genre: String? = null): Flow<List<Story>> {
        return database.storyDao().getEditorPicks(limit, genre)
            .map { entities -> entities.map { it.toModel() } }
            .onStart {
                backgroundScope.launch {
                    // TODO: Add getEditorPicks to GraphQLService
                    // graphQLService.getEditorPicks(limit, genre).fold(...)
                }
            }
    }
    
    fun getGenreStories(genre: String, page: Int = 1, limit: Int = 20): Flow<List<Story>> {
        return database.storyDao().getStoriesByGenre(genre, limit, (page - 1) * limit)
            .map { entities -> entities.map { it.toModel() } }
            .onStart {
                backgroundScope.launch {
                    // TODO: Add getGenreStories to GraphQLService
                    // graphQLService.getGenreStories(genre, page, limit).fold(...)
                }
            }
    }
}

enum class RankingType {
    MONTHLY_VOTES,
    SALES,
    RECOMMENDATIONS,
    POPULARITY
}
